### ISSUE TITLE
Lambda Functions not getting triggered

### ISSUE THREAD

**Issue by Dhariaj** [capitalone/cloud-custodian#2425]
Hello,

I am new to Cloud custodian. I have it configured and working with cross account access. I can get any policy that runs locally to work fine without any issues but when I try a policy, such as detect root login, it deploys lambda function but the function never gets triggered. I have logged in as root multiple times but still cant get the lambda function to trigger.

**Comment by paulnava** [comment]
@Dhariaj What are the permissions on the role you set for your lambda policy? 

https://github.com/capitalone/cloud-custodian/blob/master/docs/source/policy/lambda.rst#execution-options

**Comment by Dhariaj** [comment]
Lambda has full admin rights at this moment.  To clarify CC is running in account 1, If I run this policy in the same account then it creates a new lambda function and everything works well. if I run this policy using --assume and passing in a different cross account role. It creates the lambda function, associates the right role with the function but the function just sits there doesn't get triggered at all.

**Comment by thisisshi** [comment]
Do you have an event bus set up to allow events to trigger from Account 2 into Account 1? [docs](https://aws.amazon.com/blogs/aws/new-cross-account-delivery-of-cloudwatch-events/)

I haven't tried this before, but it seems like you need to set up this before you deploy the lambda into account 2

**Comment by Dhariaj** [comment]
Lambda function runs into account 2 monitoring cloudwatch in account 2. It has nothing to do with Account 1

**Comment by thisisshi** [comment]
yeah that _should_ work, is the lambda function configured to listen to the appropriate cloudwatch event in the console?

**Comment by Dhariaj** [comment]
Yes and I confirmed the same by launching the same policy in the other account and cross checking both the functions.

Sent from my iPhone

On May 21, 2018, at 2:48 PM, Sonny <notifications@github.com<mailto:notifications@github.com>> wrote:


yeah that should work, is the lambda function configured to listen to the appropriate cloudwatch event in the console?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_capitalone_cloud-2Dcustodian_issues_2425-23issuecomment-2D390747009&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=UPvjvel5_BvcNSs4ZxkQJe8EDRrrbw_56DoHm7bhBH0&s=tPcy5ghqALpZLuFQcaWTuzyOgin9HvV1Zsm8GzrBEPk&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AklEXKyoEu1kFenOa5Qf0xMbf2RR1biPks5t0ww6gaJpZM4UHFzq&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=UPvjvel5_BvcNSs4ZxkQJe8EDRrrbw_56DoHm7bhBH0&s=jvielATFAexP4NhDC6eLusjRv6GdIr1EifhdH-3ZmUQ&e=>.



Email transmitted across the Internet is normally not protected and may be intercepted and viewed by others. Therefore, you should refrain from sending any confidential or private information via unsecured email to PenFed. We will not ask you to send confidential information to us via email, such as your logon ID, password, account numbers, or Social Security number. We prohibit our employees from sending confidential information to you via email that is not encrypted. The recommended document submission method is FAX; a partial list of generic fax numbers can be found <https://www.penfed.org/aboutUs/contactUs.asp#fax> here<https://www.penfed.org/aboutUs/contactUs.asp#fax>.<https://www.penfed.org/aboutUs/contactUs.asp#fax>

**Comment by kapilt** [comment]
what is your policy that your referencing to (pls sanitize)?

**Comment by Dhariaj** [comment]
The sample policy to detect root logins. I took it directly from Capital One github page.

Sent from my iPhone

On May 21, 2018, at 7:43 PM, Kapil Thangavelu <notifications@github.com<mailto:notifications@github.com>> wrote:


what is your policy that your referencing to (pls sanitize)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_capitalone_cloud-2Dcustodian_issues_2425-23issuecomment-2D390818267&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=ke3XpAPdcf-4yeEOS8JOslQbtYFsvkI0LhqmlEcXit0&s=Eu_1egaQozcG0qR07eIYoU1HedM8cq997pFc4mIson0&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AklEXAF2APKuAT-2DQQcX5eJhXswBsz730ks5t01FZgaJpZM4UHFzq&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=ke3XpAPdcf-4yeEOS8JOslQbtYFsvkI0LhqmlEcXit0&s=px04E6H4MzwkgYr3tfq8NhHfeD9WghwKunrL9R_KCgM&e=>.



Email transmitted across the Internet is normally not protected and may be intercepted and viewed by others. Therefore, you should refrain from sending any confidential or private information via unsecured email to PenFed. We will not ask you to send confidential information to us via email, such as your logon ID, password, account numbers, or Social Security number. We prohibit our employees from sending confidential information to you via email that is not encrypted. The recommended document submission method is FAX; a partial list of generic fax numbers can be found <https://www.penfed.org/aboutUs/contactUs.asp#fax> here<https://www.penfed.org/aboutUs/contactUs.asp#fax>.<https://www.penfed.org/aboutUs/contactUs.asp#fax>

**Comment by kapilt** [comment]
whats in the lambda policy logs, is it firing, is the cloud watch event target firing?

**Comment by Dhariaj** [comment]
Lambda is not firing and the Logs were empty. Cloud watch I can see events for security group ingress and egress.

Just to confirm so that we are on the same page. The ec2 instance running cloud custodian is running in account 1. I deployed the same policy in account 1 and passed a role that lambda could assume. This role was defined in account 1. Everything works as expected.  Now I used the same policy but used a different role for lambda. This role is defined in account 2. I ran the policy using the —assume role tag and a new lambda function got deployed in account 2.  I created a new security group in account 2 and made it insecure but that didn’t trigger the lambda function. I made a bunch of changes to the SG but lambda function would not trigger.

That lambda function In account 2 has full admin rights. It also has cross account admin rights to account 1 just in case.

Jimmy
Sent from my iPhone

On May 21, 2018, at 9:06 PM, Kapil Thangavelu <notifications@github.com<mailto:notifications@github.com>> wrote:


whats in the lambda policy logs, is it firing, is the cloud watch event target firing?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_capitalone_cloud-2Dcustodian_issues_2425-23issuecomment-2D390830841&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=jrEk1m3DYx3W4nlqjyIsT1ewkcsiEnFrQ0Straiun1Q&s=5kQwtCjrEjcxgqFQpdjBglssDyu1ADXY5dnSt0monhs&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AklEXKJ-5FUSX7OBVw5SL2rGo7xY42O2rsks5t02THgaJpZM4UHFzq&d=DwMCaQ&c=uUZbzTZZgO9xXVRZz77NpQ&r=7JWmtnrZy4WGjKQ01tyHzY4bGM7dWpE-dowS5b7_B20&m=jrEk1m3DYx3W4nlqjyIsT1ewkcsiEnFrQ0Straiun1Q&s=-9Jgqb31EbE-jCsoDH_DsAiGO4T_khwi8_CokY39L38&e=>.



Email transmitted across the Internet is normally not protected and may be intercepted and viewed by others. Therefore, you should refrain from sending any confidential or private information via unsecured email to PenFed. We will not ask you to send confidential information to us via email, such as your logon ID, password, account numbers, or Social Security number. We prohibit our employees from sending confidential information to you via email that is not encrypted. The recommended document submission method is FAX; a partial list of generic fax numbers can be found <https://www.penfed.org/aboutUs/contactUs.asp#fax> here<https://www.penfed.org/aboutUs/contactUs.asp#fax>.<https://www.penfed.org/aboutUs/contactUs.asp#fax>

**Comment by kapilt** [comment]
okay..

- sounds like we dropped the root login case since that was used as proxy example for the actual use case around security group policies. 
- your policy lambdas work fine in account 1
- your policy lambdas are not working in account 2. more specifically they are not being invoked.

can you look at the metrics for the event targets in cloud watch events for account 2 to see if they are firing?

It sounds like some sort of setup issue, but its hard to see what that might be from the limited information being given (run logs, policy, metrics, versions etc).  if you want to setup some time to debug, reach out to me on gitter / @kapilt

also what version of custodian? and what version of python? and what version of linux on the ami?

**Comment by kapilt** [comment]
`custodian version --debug` has some information that would be useful to capture for this issue.

**Comment by Dhariaj** [comment]
Please copy/paste the following info along with any bug reports:

Custodian:   0.8.28.2
Python:      2.7.14 (default, May  2 2018, 18:31:34)
             [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
Platform:    ('Linux', 'ip-10-160-32-136', '4.14.33-51.37.amzn1.x86_64', '#1 SMP Thu May 3 20:07:43 UTC 2018', 'x86_64')
Using venv:  True
PYTHONPATH:
[            '/home/ec2-user/custodian/bin',
             '/home/ec2-user/custodian/local/lib64/python2.7/site-packages',
             '/home/ec2-user/custodian/local/lib/python2.7/site-packages',
             '/home/ec2-user/custodian/lib64/python2.7',
             '/home/ec2-user/custodian/lib/python2.7',
             '/home/ec2-user/custodian/lib64/python2.7/site-packages',
             '/home/ec2-user/custodian/lib/python2.7/site-packages',
             '/home/ec2-user/custodian/lib64/python2.7/lib-dynload',
             '/usr/lib64/python2.7',
             '/usr/lib/python2.7']

I invoked the policy using the following command
custodian run --output-dir=. --cache-period=10 --assume="arn:aws:iam::XXXXXX:role/CloudCustodian" -s s3://XXX/accounts/XX/XX/policies root.yml

Event from cloud trail





2018-05-21, 01:24:59 PM

xx

AuthorizeSecurityGroupIngress

EC2 SecurityGroup

sg-xx


2018-05-21, 01:24:59 PM

xx

RevokeSecurityGroupIngress

EC2 SecurityGroup

sg-xx


Thank you

Dharia, Jimmy

**Comment by tjstansell** [comment]
I ran into lambda functions not being invoked too in one of our accounts and it turns out that we didn't have a cloudtrail trail configured in that account yet.  That's required to for lambdas to receive events.  After creating a trail, I also had to go remove and re-deploy the lambdas to get them plumbed up properly and actually start receiving alerts.
